frame/bridge/eth/linear-relay/src/lib.rs

```diff
@@ -0,0 +1,988 @@
+
+	type Call: Dispatchable + From<Call<Self>> + IsSubType<Module<Self>, Self> + Clone;
+
+	type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>
+		+ ReservableCurrency<Self::AccountId>;
+}
+
+#[derive(Clone, PartialEq, Encode, Decode)]
+pub enum EthNetworkType {
+	Mainnet,
+	Ropsten,
+}
+
+impl Default for EthNetworkType {
+	fn default() -> EthNetworkType {
+		EthNetworkType::Mainnet
+	}
+}
+
+#[cfg(feature = "std")]
+darwinia_support::impl_genesis! {
+	struct DagsMerkleRootsLoader {
+		dags_merkle_roots: Vec<H128>
+	}
+}
+
+#[derive(Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+pub struct DoubleNodeWithMerkleProof {
+	pub dag_nodes: [H512; 2],
+	pub proof: Vec<H128>,
+}
+
+impl Default for DoubleNodeWithMerkleProof {
+	fn default() -> DoubleNodeWithMerkleProof {
+		DoubleNodeWithMerkleProof {
+			dag_nodes: <[H512; 2]>::default(),
+			proof: Vec::new(),
+		}
+	}
+}
+
+impl DoubleNodeWithMerkleProof {
+	pub fn from_str_unchecked(s: &str) -> Self {
+		let mut dag_nodes: Vec<H512> = Vec::new();
+		let mut proof: Vec<H128> = Vec::new();
+		for e in s.splitn(60, '"') {
+			let l = e.len();
+			if l == 34 {
+				proof.push(fixed_hex_bytes_unchecked!(e, 16).into());
+			} else if l == 130 {
+				dag_nodes.push(fixed_hex_bytes_unchecked!(e, 64).into());
+			} else if l > 34 {
+				// should not be here
+				panic!("the proofs are longer than 25");
+			}
+		}
+		DoubleNodeWithMerkleProof {
+			dag_nodes: [dag_nodes[0], dag_nodes[1]],
+			proof,
+		}
+	}
+}
+
+impl DoubleNodeWithMerkleProof {
+	pub fn apply_merkle_proof(&self, index: u64) -> H128 {
+		fn hash_h128(l: H128, r: H128) -> H128 {
+			let mut data = [0u8; 64];
+			data[16..32].copy_from_slice(&(l.0));
+			data[48..64].copy_from_slice(&(r.0));
+
+			// `H256` is 32 length, truncate is safe; qed
+			array_unchecked!(sha2_256(&data), 16, 16).into()
+		}
+
+		let mut data = [0u8; 128];
+		data[..64].copy_from_slice(&(self.dag_nodes[0].0));
+		data[64..].copy_from_slice(&(self.dag_nodes[1].0));
+
+		// `H256` is 32 length, truncate is safe; qed
+		let mut leaf = array_unchecked!(sha2_256(&data), 16, 16).into();
+		for i in 0..self.proof.len() {
+			if (index >> i as u64) % 2 == 0 {
+				leaf = hash_h128(leaf, self.proof[i]);
+			} else {
+				leaf = hash_h128(self.proof[i], leaf);
+			}
+		}
+
+		leaf
+	}
+}
+
+/// Familial details concerning a block
+#[derive(Clone, Default, PartialEq, Encode, Decode)]
+pub struct EthHeaderBrief<AccountId> {
+	/// Total difficulty of the block and all its parents
+	pub total_difficulty: U256,
+	/// Parent hash of the header
+	pub parent_hash: H256,
+	/// Block number
+	pub number: EthBlockNumber,
+	/// Relayer of the block header
+	pub relayer: AccountId,
+}
+
+#[derive(Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+pub struct EthReceiptProof {
+	pub index: u64,
+	pub proof: Vec<u8>,
+	pub header_hash: H256,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as DarwiniaEthLinearRelay {
```

Comment:
The Crab Networking are using ```DarwiniaEthRelay``` as storage key, we must be careful here when do runtime upgrading on Crab Newtork, storage migration might needed.

It could be danger is new DarwiniaEthRelay are going to use ```DarwiniaEthRelay``` too.

Reply:
I add a todo there, the migration operation will be added later.